// minio_service.proto - MinIO 对象存储服务 gRPC 接口
// 提供统一的对象存储访问接口，支持多租户和用户隔离
syntax = "proto3";

package isa.minio;

option go_package = "github.com/isa-cloud/isa_cloud/api/proto/minio";

import "common.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ========================================
// MinIO 对象存储服务
// ========================================
// 这是一个通用的对象存储服务，为所有后端服务提供文件存储能力
// 支持：
// - 用户和组织级别的隔离
// - 桶（Bucket）和对象（Object）管理
// - 流式上传/下载（支持大文件）
// - 预签名 URL（临时访问链接）
// - 对象元数据管理

service MinIOService {
    // ========== 桶管理 ==========
    rpc CreateBucket(CreateBucketRequest) returns (CreateBucketResponse);
    rpc ListBuckets(ListBucketsRequest) returns (ListBucketsResponse);
    rpc DeleteBucket(DeleteBucketRequest) returns (DeleteBucketResponse);
    rpc GetBucketInfo(GetBucketInfoRequest) returns (GetBucketInfoResponse);
    rpc SetBucketPolicy(SetBucketPolicyRequest) returns (SetBucketPolicyResponse);
    rpc GetBucketPolicy(GetBucketPolicyRequest) returns (GetBucketPolicyResponse);
    
    // ========== 对象操作 ==========
    // 流式上传（支持大文件分块上传）
    rpc PutObject(stream PutObjectRequest) returns (PutObjectResponse);
    // 流式下载（支持大文件分块下载）
    rpc GetObject(GetObjectRequest) returns (stream GetObjectResponse);
    // 删除对象
    rpc DeleteObject(DeleteObjectRequest) returns (DeleteObjectResponse);
    // 批量删除对象
    rpc DeleteObjects(DeleteObjectsRequest) returns (DeleteObjectsResponse);
    // 列出对象
    rpc ListObjects(ListObjectsRequest) returns (ListObjectsResponse);
    // 复制对象
    rpc CopyObject(CopyObjectRequest) returns (CopyObjectResponse);
    // 获取对象信息（不下载内容）
    rpc StatObject(StatObjectRequest) returns (StatObjectResponse);
    
    // ========== 预签名 URL ==========
    rpc GetPresignedURL(GetPresignedURLRequest) returns (GetPresignedURLResponse);
    rpc GetPresignedPutURL(GetPresignedPutURLRequest) returns (GetPresignedPutURLResponse);
    
    // ========== 多部分上传 ==========
    rpc InitiateMultipartUpload(InitiateMultipartUploadRequest) returns (InitiateMultipartUploadResponse);
    rpc UploadPart(stream UploadPartRequest) returns (UploadPartResponse);
    rpc CompleteMultipartUpload(CompleteMultipartUploadRequest) returns (CompleteMultipartUploadResponse);
    rpc AbortMultipartUpload(AbortMultipartUploadRequest) returns (AbortMultipartUploadResponse);
    
    // ========== 健康检查 ==========
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// ========================================
// 数据结构
// ========================================

// 桶信息
message BucketInfo {
    string name = 1;
    google.protobuf.Timestamp creation_date = 2;
    string owner_id = 3;                          // 桶所有者用户ID
    string organization_id = 4;                   // 所属组织ID
    map<string, string> tags = 5;                 // 桶标签
    string region = 6;                            // 区域
    int64 size_bytes = 7;                         // 桶总大小（字节）
    int64 object_count = 8;                       // 对象数量
}

// 对象信息
message ObjectInfo {
    string key = 1;                               // 对象键（路径）
    int64 size = 2;                               // 大小（字节）
    string etag = 3;                              // ETag（内容哈希）
    string content_type = 4;                      // MIME 类型
    google.protobuf.Timestamp last_modified = 5;  // 最后修改时间
    map<string, string> metadata = 6;             // 自定义元数据
    string storage_class = 7;                     // 存储类型
    string version_id = 8;                        // 版本ID（如果启用版本控制）
    string owner_id = 9;                          // 对象所有者
}

// 桶策略类型
enum BucketPolicyType {
    BUCKET_POLICY_PRIVATE = 0;      // 私有（默认）
    BUCKET_POLICY_PUBLIC_READ = 1;  // 公开读
    BUCKET_POLICY_PUBLIC_WRITE = 2; // 公开写
    BUCKET_POLICY_PUBLIC_RW = 3;    // 公开读写
    BUCKET_POLICY_CUSTOM = 4;       // 自定义策略
}

// ========================================
// 桶管理 - 请求/响应
// ========================================

message CreateBucketRequest {
    string bucket_name = 1;
    string user_id = 2;
    string organization_id = 3;
    string region = 4;
    bool object_locking = 5;                      // 是否启用对象锁定
    map<string, string> tags = 6;
}

message CreateBucketResponse {
    bool success = 1;
    string message = 2;
    BucketInfo bucket_info = 3;
    string error = 4;
}

message ListBucketsRequest {
    string user_id = 1;
    string organization_id = 2;
    string prefix = 3;                            // 桶名前缀过滤
}

message ListBucketsResponse {
    bool success = 1;
    repeated BucketInfo buckets = 2;
    int32 total_count = 3;
    string error = 4;
}

message DeleteBucketRequest {
    string bucket_name = 1;
    string user_id = 2;
    bool force = 3;                               // 强制删除（包括所有对象）
}

message DeleteBucketResponse {
    bool success = 1;
    string message = 2;
    int32 deleted_objects = 3;                    // 删除的对象数量
    string error = 4;
}

message GetBucketInfoRequest {
    string bucket_name = 1;
    string user_id = 2;
}

message GetBucketInfoResponse {
    bool success = 1;
    BucketInfo bucket_info = 2;
    string error = 3;
}

message SetBucketPolicyRequest {
    string bucket_name = 1;
    string user_id = 2;
    BucketPolicyType policy_type = 3;
    string custom_policy = 4;                     // JSON 格式的自定义策略
}

message SetBucketPolicyResponse {
    bool success = 1;
    string message = 2;
    string error = 3;
}

message GetBucketPolicyRequest {
    string bucket_name = 1;
    string user_id = 2;
}

message GetBucketPolicyResponse {
    bool success = 1;
    BucketPolicyType policy_type = 2;
    string policy_json = 3;
    string error = 4;
}

// ========================================
// 对象操作 - 请求/响应
// ========================================

// 上传对象（流式）
message PutObjectRequest {
    oneof data {
        PutObjectMetadata metadata = 1;           // 第一个消息：元数据
        bytes chunk = 2;                          // 后续消息：数据块
    }
}

message PutObjectMetadata {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string content_type = 4;
    int64 content_length = 5;                     // 总大小
    map<string, string> metadata = 6;
    string storage_class = 7;
}

message PutObjectResponse {
    bool success = 1;
    string object_key = 2;
    string etag = 3;
    string version_id = 4;
    int64 size = 5;
    string error = 6;
}

// 下载对象（流式）
message GetObjectRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    int64 offset = 4;                             // 字节偏移量（支持断点续传）
    int64 length = 5;                             // 读取长度（0表示读取全部）
    string version_id = 6;                        // 版本ID（可选）
}

message GetObjectResponse {
    oneof data {
        ObjectInfo metadata = 1;                  // 第一个消息：对象信息
        bytes chunk = 2;                          // 后续消息：数据块
    }
}

// 删除对象
message DeleteObjectRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string version_id = 4;                        // 版本ID（可选）
}

message DeleteObjectResponse {
    bool success = 1;
    string message = 2;
    string error = 3;
}

// 批量删除对象
message DeleteObjectsRequest {
    string bucket_name = 1;
    string user_id = 2;
    repeated string object_keys = 3;
    bool quiet = 4;                               // 静默模式（只返回错误）
}

message DeleteObjectsResponse {
    bool success = 1;
    repeated string deleted_keys = 2;
    repeated DeleteError errors = 3;
    string error = 4;
}

message DeleteError {
    string key = 1;
    string error_code = 2;
    string error_message = 3;
}

// 列出对象
message ListObjectsRequest {
    string bucket_name = 1;
    string user_id = 2;
    string prefix = 3;                            // 前缀过滤
    string delimiter = 4;                         // 分隔符（用于模拟文件夹）
    int32 max_keys = 5;                           // 最大返回数量
    string start_after = 6;                       // 从哪个key之后开始
    string continuation_token = 7;                // 分页令牌
    bool recursive = 8;                           // 是否递归列出
}

message ListObjectsResponse {
    bool success = 1;
    repeated ObjectInfo objects = 2;
    repeated string common_prefixes = 3;          // 公共前缀（文件夹）
    string next_continuation_token = 4;
    bool is_truncated = 5;                        // 是否还有更多
    string error = 6;
}

// 复制对象
message CopyObjectRequest {
    string source_bucket = 1;
    string source_key = 2;
    string dest_bucket = 3;
    string dest_key = 4;
    string user_id = 5;
    map<string, string> metadata = 6;
    string metadata_directive = 7;                // COPY 或 REPLACE
}

message CopyObjectResponse {
    bool success = 1;
    string message = 2;
    string etag = 3;
    google.protobuf.Timestamp last_modified = 4;
    string error = 5;
}

// 获取对象信息
message StatObjectRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string version_id = 4;
}

message StatObjectResponse {
    bool success = 1;
    ObjectInfo object_info = 2;
    string error = 3;
}

// ========================================
// 预签名 URL - 请求/响应
// ========================================

message GetPresignedURLRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    int32 expiry_seconds = 4;                     // URL 过期时间（秒，默认7天）
    map<string, string> request_params = 5;       // 额外的请求参数
}

message GetPresignedURLResponse {
    bool success = 1;
    string url = 2;
    google.protobuf.Timestamp expires_at = 3;
    string error = 4;
}

message GetPresignedPutURLRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    int32 expiry_seconds = 4;
    string content_type = 5;
}

message GetPresignedPutURLResponse {
    bool success = 1;
    string url = 2;
    google.protobuf.Timestamp expires_at = 3;
    string error = 4;
}

// ========================================
// 多部分上传 - 请求/响应
// ========================================

message InitiateMultipartUploadRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string content_type = 4;
    map<string, string> metadata = 5;
}

message InitiateMultipartUploadResponse {
    bool success = 1;
    string upload_id = 2;
    string bucket_name = 3;
    string object_key = 4;
    string error = 5;
}

message UploadPartRequest {
    oneof data {
        UploadPartMetadata metadata = 1;
        bytes chunk = 2;
    }
}

message UploadPartMetadata {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string upload_id = 4;
    int32 part_number = 5;                        // 分片编号（从1开始）
    int64 part_size = 6;                          // 分片大小
}

message UploadPartResponse {
    bool success = 1;
    int32 part_number = 2;
    string etag = 3;
    string error = 4;
}

message CompletedPart {
    int32 part_number = 1;
    string etag = 2;
}

message CompleteMultipartUploadRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string upload_id = 4;
    repeated CompletedPart parts = 5;
}

message CompleteMultipartUploadResponse {
    bool success = 1;
    string object_key = 2;
    string etag = 3;
    string location = 4;
    string error = 5;
}

message AbortMultipartUploadRequest {
    string bucket_name = 1;
    string object_key = 2;
    string user_id = 3;
    string upload_id = 4;
}

message AbortMultipartUploadResponse {
    bool success = 1;
    string message = 2;
    string error = 3;
}

// ========================================
// 健康检查
// ========================================

message HealthCheckRequest {
    bool detailed = 1;                            // 是否返回详细信息
}

message HealthCheckResponse {
    bool success = 1;
    bool healthy = 2;
    string status = 3;
    google.protobuf.Timestamp timestamp = 4;
    google.protobuf.Struct details = 5;           // 详细信息（连接池、延迟等）
    string error = 6;
}



